### PR TITLE
Expose query accesses

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

It would be useful to be able to inspect a `QueryState`'s accesses so we can detect when the data it accesses changes without having to iterate it. However there are two things preventing this:

* These accesses are unnecessarily encapsulated.
* `Has<T>` indirectly accesses `T`, but does not register it.

## Solution

* Expose accesses and matches used by `QueryState`.
* Add the notion of "archetypal" accesses, which are not accessed directly, but whose presence in an archetype affects a query result.

### CODE REVIEW

File: crates/bevy_ecs/src/query/access.rs [hunk 1]
```diff
@@ -116,6 +119,12 @@ impl<T: SparseSetIndex> Access<T> {
         self.writes.insert(index.sparse_set_index());
     }
 
+    /// Adds an archetypal (inderect) access to the element given by `index`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:122]:
```suggestion
    /// Adds an archetypal (indirect) access to the element given by `index`.
```

Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:122]: These docs should capture the same information as `has_archetypal`

File: crates/bevy_ecs/src/query/access.rs [hunk 2]
```diff
@@ -272,6 +289,14 @@ impl<T: SparseSetIndex> Access<T> {
     pub fn writes(&self) -> impl Iterator<Item = T> + '_ {
         self.writes.ones().map(T::get_sparse_set_index)
     }
+
+    /// Returns the indices of the elements that this has an archetypal access to.
+    ///
+    /// Archetypal accesses will never lead to conflicts, but the presence of the data
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:295]: Explanatory docs should be synced up here too.

File: crates/bevy_ecs/src/query/access.rs [hunk 3]
```diff
@@ -469,6 +494,20 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     pub fn is_subset(&self, other: &FilteredAccess<T>) -> bool {
         self.required.is_subset(&other.required) && self.access().is_subset(other.access())
     }
+
+    /// Returns the components this access filters for.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:498]: Naming and docs could be clearer here and below.

File: crates/bevy_ecs/src/query/state.rs [hunk 4]
```diff
@@ -95,6 +95,26 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
     ) -> &QueryState<NewD, NewF> {
         &*(self as *const QueryState<D, F> as *const QueryState<NewD, NewF>)
     }
+
+    /// Returns the archetype components accessed by this query.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:99]:
```suggestion
    /// Returns the archetype-components accessed by this query.
```

**Review by alice-i-cecile — COMMENTED** [review]
And so we inch closer to the access model laid out in https://github.com/bevyengine/rfcs/pull/30 :D

I agree with this direction, and this PR is useful both as an ECS primitive and as something we can expose to users. I'd like to clean up the explanation and docs a bit further before merging.

Ideally we'd have module level documentation explaining the different kinds of access and how they relate to each other, but I won't block on that in this PR.

File: crates/bevy_ecs/src/query/state.rs (hunk 4 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/state.rs:99]: Hmm, "archetype component" is a compound noun which I feel usually reads better without a hyphen. I prefer to save hyphens for compound adjectives

File: crates/bevy_ecs/src/query/access.rs (hunk 3 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/access.rs]: Let me know if it's still not clear enough

File: crates/bevy_ecs/src/query/access.rs [hunk 5]
```diff
@@ -116,6 +119,15 @@ impl<T: SparseSetIndex> Access<T> {
         self.writes.insert(index.sparse_set_index());
     }
 
+    /// Adds an archetypal (indirect) access to the element given by `index`.
+    ///
+    /// This is for elements that are not accessed (and thus will never cause conflicts),
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:124]:
```suggestion
    /// This is for elements whose values are not accessed (and thus will never cause conflicts),
```

File: crates/bevy_ecs/src/query/access.rs [hunk 6]
```diff
@@ -136,6 +148,14 @@ impl<T: SparseSetIndex> Access<T> {
         self.writes_all || !self.writes.is_clear()
     }
 
+    /// Returns true if this has an archetypal (indirect) access to the element given by `index`.
+    ///
+    /// This is an element that is not accessed (and thus will never cause conflicts),
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:153]:
```suggestion
    /// This is an element whose value is not accessed (and thus will never cause conflicts),
```

File: crates/bevy_ecs/src/query/access.rs [hunk 7]
```diff
@@ -272,6 +292,14 @@ impl<T: SparseSetIndex> Access<T> {
     pub fn writes(&self) -> impl Iterator<Item = T> + '_ {
         self.writes.ones().map(T::get_sparse_set_index)
     }
+
+    /// Returns the indices of the elements that this has an archetypal access to.
+    ///
+    /// These are elements that are not accessed (and thus will never cause conflicts),
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:298]:
```suggestion
    /// These are elements whose values are not accessed (and thus will never cause conflicts),
```

**Review by alice-i-cecile — APPROVED** [review]
Solid stuff. One small set of suggestions to clarify further, but I'm happy with this now.

**Comment by JoJoJet** [comment]
> Ideally we'd have module level documentation explaining the different kinds of access and how they relate to each other, but I won't block on that in this PR.

I'll think on what I would write here, but I probably won't include anything in this PR since module-level documentation generally requires a lot more thought than method-level documentation.

File: crates/bevy_ecs/src/query/access.rs [hunk 8]
```diff
@@ -59,6 +59,8 @@ pub struct Access<T: SparseSetIndex> {
     /// Is `true` if this has mutable access to all elements in the collection.
     /// If this is true, then `reads_all` must also be true.
     writes_all: bool,
+    // Elements that are not accessed, but whose presence in an archetype affect query results.
+    archetypal: FixedBitSet,
```
Comment (james7132) [inline crates/bevy_ecs/src/query/access.rs:63]: Shouldn't these be populated by the appropriate QueryData/QueryFilter/SystemParam implementors? or is the plan to handle that in a separate PR?

Comment (JoJoJet) [reply crates/bevy_ecs/src/query/access.rs:63]: Oh right, forgot to move that over from my branch.

**Review by james7132 — APPROVED** [review]

**Pull request merged**